model-optimizer/mo/back/preprocessing.py

```diff
@@ -0,0 +1,401 @@
+    # On return, mean_scale_values is a dictionary with input names as key and mean/scale pair as value
+    # {'inputName': {'mean': [1., 2., 3.], 'scale': [2.]}}
+
+    layout_values = {}
+    if 'layout_values' in argv and argv.layout_values:
+        layout_values = argv.layout_values
+
+    if '' in layout_values:
+        if len(ov_function.inputs) > 1:
+            input_names = [list(ov_input.get_tensor().get_names())[0] for ov_input in ov_function.inputs]
+            raise Error('Layout without name can be specified for models with only one input, '
+                        'but provided model has {} inputs: \'{}\'. '
+                        'Please specify explicitly input/output name for --layout option'
+                        .format(len(input_names), input_names))
+        layout_values = {
+            list(ov_function.input().get_tensor().get_names())[0]: {
+                'source_layout': layout_values[''].get('source_layout'),
+                'target_layout': layout_values[''].get('target_layout')
+            }
+        }
+    check_keys_valid(ov_function=ov_function, keys=mean_scale_values.keys(), search_outputs=False)
+    check_keys_valid(ov_function=ov_function, keys=layout_values.keys(), search_outputs=True)
+
+    layout_values = update_layout_is_input_flag(ov_function, layout_values)
+    layout_values = guess_source_layouts_by_mean_scale(ov_function, layout_values, mean_scale_values)
+    need_reverse = 'reverse_input_channels' in argv and argv.reverse_input_channels
+    suitable_params_ric = []
+    if need_reverse:
+        suitable_params_ric = guess_source_layouts_for_reverse_channels(ov_function=ov_function,
+                                                                        layout_values=layout_values)
+
+    for node_name, layout_value in layout_values.items():
+        if layout_value.get('source_layout'):
+            if layout_value.get('is_input'):
+                prep.input(node_name).network().set_layout(Layout(layout_value['source_layout']))
+            else:
+                prep.output(node_name).network().set_layout(Layout(layout_value['source_layout']))
+        if layout_value.get('target_layout'):
+            if layout_value.get('is_input'):
+                prep.input(node_name).tensor().set_layout(Layout(layout_value['target_layout']))
+            else:
+                prep.output(node_name).tensor().set_layout(Layout(layout_value['target_layout']))
+
+    for node_name, node_mean_scale_values in mean_scale_values.items():
+        # Apply mean first, then scale
+        if node_mean_scale_values['mean'] is not None:
+            prep.input(node_name).preprocess().mean(node_mean_scale_values['mean'])
+        if node_mean_scale_values['scale'] is not None:
+            prep.input(node_name).preprocess().scale(node_mean_scale_values['scale'])
+        log.debug('Mean/Scale pre-processing applied to {}'.format(node_name))
+
+    # Apply reverse_input_channels
+    if need_reverse:
```

Comment:
I think reverse_input_channels should be applied before mean/scale

Reply:
MO help says different thing:
```
    common_group.add_argument('--reverse_input_channels',
                              help='Switch the input channels order from RGB to BGR (or vice versa). Applied to '
                                   'original inputs of the model if and only if a number of channels equals 3. Applied '
                                   'after application of --mean_values and --scale_values options, so numbers in '
                                   '--mean_values and --scale_values go in the order of channels used in the original '
                                   'model.',
                              action='store_true')
```